Simplify & rectify dequantized B buffer loading for AMX GEMM micro-kernel for WoQ int8 case

sanchitintel opened the pull request:
As suggested by @leslie-fang-intel in https://github.com/leslie-fang-intel/pytorch/commit/4c83e4e75138e8fa6e0d58438f75b7718dc8a0cc#diff-139642bd981df977f70f4c18c1c34bd1a85c1d6b9ffa06aaa98426ed83942a31R537 - all elements of `B` tiles (not referring to AMX tiles, but the tiles at the granularity of the micro-kernel) have contiguous elements since `B` matrix is pre-packed, so dequantized buffer loading logic can be simplified. While the previous approach kept elements to be loaded into a B AMX tile contiguous, the new approach doesn't entail any performance penalty either because that data is already in L1D, so loading AMX tiles from non-contiguous dequantized B elements doesn't adversely affect performance.

Also rectified the size of the dequantized B buffer.

Fixes #140208.

A subsequent PR will factor out caching of dequantized int8 weights into a separate codegen function


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

sanchitintel committed:
Simplify B tile loading logic for AMX GEMM micro-kernel

As suggested by @leslie-fang-intel in https://github.com/leslie-fang-intel in /pytorch/commit/4c83e4e75138e8fa6e0d58438f75b7718dc8a0cc#diff-139642bd981df977f70f4c18c1c34bd1a85c1d6b9ffa06aaa98426ed83942a31R537

sanchitintel committed:
Clarify comment

sanchitintel committed:
Support N != block_n case

This case cannot be covered by the current UTs, since it hasn't been implemented

sanchitintel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,45 +528,25 @@ class CppMicroGemmAMX(CppMicroGemm):
-                    num_b_rows
-                );
-            }
+        // The assumption here is that if N would be a multiple of block_n,
+        // then B would be some consecutive tiles sized [k, block_n] each, so the
+        // first element of a subsequent [K, block_n] tile would be right after the last
+        // element of the previous tile sized [K x block_n]
```
Hi @jgong5, can you please confirm if this assumption is true? Thanks!

Since the `N != block_n` case can't be currently tested, I'm not sure if this is what you have in mind.

Also, as a digression, please note that `N != block_n` case would require loading `A` tiles multiple times.

jgong5 replied:
No, please check my other comment.

sanchitintel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,45 +528,25 @@ class CppMicroGemmAMX(CppMicroGemm):
+        // first element of a subsequent [K, block_n] tile would be right after the last
+        // element of the previous tile sized [K x block_n]
+        auto base_idx = n * K * {{block_n}};
+        for (int idx = 0; idx < buf_size; idx += {{block_k}}) {
```
@leslie-fang-intel, since `block_k` is always 32 for int8 case, I removed the `if-else` conditional in the patch.

leslie-fang-intel replied:
- I still feels better to use the 32 here which seems more readable.
- What's the case when `buf_size (K * {{block_n}})` is not divisible by `block_k(32)`? Are the load and store with
memory access out of bounds implicitly?

sanchitintel replied:
`K` is always a multiple of 2, and `Nr` is either 16, 32 or 48.
So, `K * Nr` is always a multiple of 32.

sanchitintel committed:
Change variable name from `n` to `tile_idx` for clarity

sanchitintel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,45 +528,25 @@ class CppMicroGemmAMX(CppMicroGemm):
-    const auto b_tile_ptr_stride = ldb * {{vnni_size}};
-
-    auto load_B_row = [&]({{input2_t}}* {{restrict_keyword}} src, {{input_t}}* {{restrict_keyword}} dst) {
-        auto b_int8 = at::vec::Vectorized<int8_t>::loadu(src, static_cast<int64_t>(32));
-        auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
-        b_bf16.store(dst);
-    };
-
-    auto load_B_tile = [&]({{input2_t}}* B_ptr, int idx, int num_b_rows) {
-        {{input_t}}* base_addr = dequantized_B_buf + idx;
-        {{kernel.unroll_pragma(8)}}
-        for (int i = 0; i < num_b_rows; i++) {
-            load_B_row(
-                B_ptr + i * b_tile_ptr_stride,
-                base_addr + i * 32
-            );
-        }
-    };
-    auto load_dequantized_B = [&](int n) {
+    auto load_dequantized_B = [&](int tile_idx) {
         // Load a tile of B & cache it in L1D.
-        {{kernel.unroll_pragma(4)}}
-        for (int k = 0; k < K; k += {{block_k}}) {
-            int num_b_rows = (k < last_k_offset) ? 16 : tail_k_size;
-            {{kernel.unroll_pragma(2)}}
-            for (int tile_col = 0; tile_col <= 1; tile_col++) {
-                load_B_tile(
```
Hi @leslie-fang-intel, the previous approach placed the elements of an AMX B tile contiguously.
I'm surprised to see that the new approach doesn't result in a performance penalty. 😮

leslie-fang-intel replied:
TBH, I haven't checked the performance of patch https://github.com/leslie-fang-intel/pytorch/commit/4c83e4e75138e8fa6e0d58438f75b7718dc8a0cc#diff-139642bd981df977f70f4c18c1c34bd1a85c1d6b9ffa06aaa98426ed83942a31R537

sanchitintel committed:
Fix lint

leslie-fang-intel commented:
BTW: I think horizontal transverse doesn't work well with this cache optimization cc @jgong5 @chunyuan-w

sanchitintel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,45 +528,28 @@ class CppMicroGemmAMX(CppMicroGemm):
-                    num_b_rows
-                );
-            }
+        // The assumption here is that if N would be a multiple of block_n,
+        // then B would be some consecutive tiles sized [k, block_n] elements each, so the
+        // first element of a subsequent [K, block_n] tile would be right after the last
+        // element of the previous tile sized [K x block_n] elements.
```
Hi, `ldb` is not required here (at least for `N == Nr`), since with this implementation, the buffer has int8 -> bf16 data from `B` in the same order as `B`.

In case of `N != Nr`, though, I guess all the `K * N` elements would be contiguous in `B`, but please confirm. Thanks!

jgong5 replied:
The API semantics is that `B` is of the shape `[K, N]` with stride `[ldb, 1]` where `ldb` is not necessarily equal to `N`. I don't think it is correct to assume `B` is contiguous.

sanchitintel replied:
> I don't think it is correct to assume B is contiguous.

Thanks! Please clarify if you mean `B` tiles may not be contiguous if `N != block_n` (which is currently not the case).

 `B` is currently contiguous due to `N` in the micro-kernel being equal to `block_n`, and weight pre-packing of the whole B matrix results in `B` tiles (used by the micro-kernel) being contiguous.

> `ldb` is not necessarily equal to `N`

Does this mean that to support `N != Nr` in the micro-kernel, weight-packing implementation would not have to be changed? Thanks!

sanchitintel replied:
The implementation (& the comment) does seem incorrect for `N != block_n`, because if a B-tile is shaped `[K, N]`, then `N != block_n` can't be supported by multiple `[K, block_n]` blocks being contiguous (at the granularity of blocks, although all the elements of the blocks would also be contiguous) in memory because that case would actually be like `[K * constant, block_n]`.

Please clarify if this is close to what you have in mind for `N != block_n` (I'm guessing it'd also require a change in the weight-packing implementation):

<img width="723" alt="image" src="https://github.com/user-attachments/assets/36f3938f-8c70-44d4-a806-f40d8d45b2f9">

sanchitintel replied:
@leslie-fang-intel, I feel like `N != block_n` would make dequantizing & caching a `[K, block_n]` block hard if `N != block_n`. That's because the each row of a `K, block_n` block would be discontiguous (the next row would not be adjacent to it).

The simplified dequantization code from your patch won't work in that case (`block_n != 32` and `block_n != N`).
So, it looks like we'd have to revert to the previous implementation (but adding another patch for #140208).

jgong5 replied:
Yes, this could be a possible scenario. Thanks.

sanchitintel replied:
Thanks, I revised for such a case

sanchitintel replied:
@leslie-fang-intel, looks like the approach to dequantize 32 elements at a time fails when `N != Nr` and `block_n` is either 16 or 48.

That's because we can't dequantize 32 elements at a time in that case, since they're not contiguous. 
We can only dequantize 16 at a time in case of `block_n` being 16, 
and 32 followed by 16 (or vice-versa) in case of `block_n` being 48

sanchitintel commented:
> BTW: I think horizontal transverse doesn't work well with this cache optimization cc @jgong5 @chunyuan-w

Hi, would the horizontal traverse strategy complement the existing GEMM template, or would it replace it? Thanks!

sanchitintel committed:
Revise as per review comments

leslie-fang-intel commented:
> Hi, would the horizontal traverse strategy complement the existing AMX GEMM micro-kernel template (by conditionally using it), or would it replace it? Thanks!

I think we will use it conditionally

sanchitintel committed:
Revise based on review comment

sanchitintel committed:
Rectify support of N != block_n case

This case can't be tested, though, as N != block_n case has not been implemented.

sanchitintel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,44 +528,41 @@ class CppMicroGemmAMX(CppMicroGemm):
+            );
+            auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
+            b_bf16.store(dequantized_B_buf + idx_dq);
+        {%- elif block_n == 48 %}
```
Hi @jgong5, please advise if there's any way to add error checking to ensure `block_n` is either 16, 32 or 48. Thanks!

leslie-fang-intel replied:
Maybe check `block_n` here to decide using `use_cached_dequantized_B` or not

https://github.com/pytorch/pytorch/blob/8b55f649610eab61ad3504bd0685da97dd95809e/torch/_inductor/codegen/cpp_micro_gemm.py#L728

leslie-fang-intel replied:
Maybe check the `block_n` in `codegen_define` when decide to use `use_cached_dequantized_B`.

jgong5 replied:
This seems over-complicated and also makes the logic limited to handle 16, 32 and 48. Can you refer to the following code (in `CppMicroGemmFP32Vec`) to simplify and generalize your implementation? Basically, there are two parts, i.e. 32-element chunks and the tail (<32).
```c++
    for (int64_t m = 0; m < M; m += {{block_m}}) {
        int64_t block_m = std::min<int64_t>(M - m, {{block_m}});
        for (int64_t n = 0; n < N; n += {{block_n}}) {
            if (block_m == {{block_m}}) {
                {{kernel_name}}_kernel<{{block_m}}, {{block_n}}, accum>(
                    A + m * lda,
                    B + n,
                    C + m * ldc + n,
                    K,
                    lda,
                    ldb,
                    ldc
                );
            } else {
                switch (block_m) {
{%- for b in range(block_m - 1, 0, -1) %}
                case {{b}}:
                    {{kernel_name}}_kernel<{{b}}, {{block_n}}, accum>(
                        A + m * lda,
                        B + n,
                        C + m * ldc + n,
                        K,
                        lda,
                        ldb,
                        ldc
                    );
                    break;
{%- endfor %}
                default:
                    {{kernel.assert_function}}(false, "Unsupported block_m: {{block_m}}");
                }
            }
        }
    }
```

sanchitintel replied:
Done, thanks!

leslie-fang-intel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,44 +528,41 @@ class CppMicroGemmAMX(CppMicroGemm):
-            }
+        // The assumption here is that if N would be a multiple of block_n.
+        {{input2_t}}* base_addr = const_cast<{{input2_t}}*>(B) + base_idx;
+        for (int idx_dq = 0, idx_q = 0; idx_dq < buf_size; idx_q += ldb, idx_dq += {{block_n}}) {
```
Thanks for the PR. I am wondering what's the case when `ldb` != `block_n`? It seems we only handle the first number of `block_n` elements.

sanchitintel replied:
Hi @leslie-fang-intel, as discussed offline, this code assumes weight-blocking as in https://github.com/pytorch/pytorch/pull/140258#discussion_r1841341702.

The code before today's commits assumed https://github.com/pytorch/pytorch/pull/140258#discussion_r1836285920.

leslie-fang-intel replied:
Thanks for the explanation. I am not sure about the scenario that `ldb` != `block_n`. To understand if this line of code ` for (int idx_dq = 0, idx_q = 0; idx_dq < buf_size; idx_q += ldb, idx_dq += {{block_n}}) {`, I guess we need to know the scenario. What do you think?

sanchitintel replied:
> It seems we only handle the first number of `block_n` elements

@leslie-fang-intel, can you please elaborate on this statement? Thanks!

leslie-fang-intel replied:
> > It seems we only handle the first number of `block_n` elements
> 
> @leslie-fang-intel, can you please elaborate on this statement? Thanks!

Maybe I didn't understand the scenario for why `idx_q` step by `ldb` but `idx_dq` step by `block_n`. Can you describe more about the scenario why the step of `idx_q` and `idx_dq` is different?

sanchitintel replied:
> I guess we need to know the scenario. What do you think?

As [discussed here](https://github.com/pytorch/pytorch/pull/140258#discussion_r1841308006), I think this case isn't currently possible due to the current weight-packing implementation.

> Can you describe more about the scenario why the step of `idx_q` and `idx_dq` is different?

This scenario requires different weight-blocking implementation, as in https://github.com/pytorch/pytorch/pull/140258#discussion_r1841341702

The `ldb` in that figure is `N` but `N != block_n`.

sanchitintel replied:
> since the weight has been packed into `[padded_n // block_n, k, block_n]`

Do you know why [this scenario](https://github.com/pytorch/pytorch/pull/140258#discussion_r1836285920) is not true? Thanks!

<img width="598" alt="image" src="https://github.com/user-attachments/assets/12270c7c-256d-479d-8b88-66ab45ab4a88">

leslie-fang-intel replied:
- We currently pack weight into `[padded_n // block_n, k, block_n]`
- We get the `ldb` of micro gemm input as `ldb = kernel.stride(B, 0)` https://github.com/pytorch/pytorch/blob/b379a28a95aa0b0d167f08b8d70c8b85a248309d/torch/_inductor/codegen/cpp_micro_gemm.py#L142

So, I think per current implementation, `block_n` should equal to `ldb`. Nevertheless, as @jgong5 mentions here for API semantics: https://github.com/pytorch/pytorch/pull/140258#discussion_r1839529082

> The API semantics is that B is of the shape [K, N] with stride [ldb, 1] where ldb is not necessarily equal to N. I don't think it is correct to assume B is contiguous.

I guess using `ldb` to `idx_q += ldb` should be fine.

sanchitintel replied:
> the micro-gemm makes any assumption that the weight is "blocked" at the upper level, right

@jgong5, yes, the weight-packing at upper level is transparent to the _current_ micro-kernel implementation, which uses `N == block_n`. 

But if `N != block_n`, then the weight-packing format seems to matter to the micro-kernel,

leslie-fang-intel replied:
Do we need to change this line if `use_cached_dequantized_B`?
https://github.com/pytorch/pytorch/blob/c79e78b5034198f9d6801b4fef710b9b9b0e9193/torch/_inductor/codegen/cpp_micro_gemm.py#L593

sanchitintel replied:
That code pertains to calling the inner kernel in the micro-kernel.
So, no change is required, because during loading of AMX tiles in the inner-kernel, the dequantized B cache would have the same `ldb` as `B`.

Basically, with this implementation, the elements of `B` have the same indices in dequantized B cache.

sanchitintel replied:
@jgong5, please help resolve a confusion - can the AMX micro-kernel be used even if the weight is not packed with `[N//block_n, K, block_n]` layout? I mean does it simply need VNNI layout without requiring weight-prepacking `[N//block_n, K, block_n]` one level above it? Thanks!

leslie-fang-intel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,44 +528,41 @@ class CppMicroGemmAMX(CppMicroGemm):
-                );
-            }
+        // The assumption here is that if N would be a multiple of block_n.
+        {{input2_t}}* base_addr = const_cast<{{input2_t}}*>(B) + base_idx;
```
This implementation which passes `n` as `base_idx` feels wrong to me:

https://github.com/pytorch/pytorch/blob/8b55f649610eab61ad3504bd0685da97dd95809e/torch/_inductor/codegen/cpp_micro_gemm.py#L563-L567

Although, current implementation can assume `N == block_n`, so it may not cause correctness issue. But what's the semantic of `base_idx`?

sanchitintel replied:
Resolving after offline discussion

leslie-fang-intel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,44 +528,41 @@ class CppMicroGemmAMX(CppMicroGemm):
+                static_cast<int64_t>({{block_n}})
+            );
+            auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
+            b_bf16.store(dequantized_B_buf + idx_dq);
```
Shouldn't be `b_bf16.store(dequantized_B_buf + idx_dq, {{block_n}});`? If `{{block_n}}` is less than full vec lane, why we store number of full vec lane elements into dst?

sanchitintel replied:
Thanks for catching this!

The opaque `at::vec::Vectorized` API is being used, and not intrinsics directly, I had wrongly assumed it'd automatically use AVX2 if 16 elements are read with `loadu`, but [you're right](https://github.com/pytorch/pytorch/blob/8b55f649610eab61ad3504bd0685da97dd95809e/aten/src/ATen/cpu/vec/vec512/vec512_int.h#L763) - `at::vec::Vectorized`  will use a ZMM register if AVX512 is supported.

sanchitintel replied:
Not required anymore

leslie-fang-intel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,44 +528,41 @@ class CppMicroGemmAMX(CppMicroGemm):
+                static_cast<int64_t>(32)
+            );
+            auto b_bf16_32 = at::vec::convert<{{input_t}}>(b_int8_32);
+            b_bf16_32.store(dequantized_B_buf + idx_dq);
```
ditto

leslie-fang-intel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,44 +528,41 @@ class CppMicroGemmAMX(CppMicroGemm):
+                static_cast<int64_t>(16)
+            );
+            auto b_bf16_16 = at::vec::convert<{{input_t}}>(b_int8_16);
+            b_bf16_16.store(dequantized_B_buf + idx_dq + 32);
```
ditto

sanchitintel committed:
Revert `load_Dequantized_B` to this PR's original implementation

sanchitintel committed:
Revise comments to clarify

sanchitintel commented:
@jgong5 @leslie-fang-intel,

For BF16xBF16 & int8xint8 GEMMs, the AMX micro-kernel would need more changes to support `N != block_n` - for instance, if `N != block_n` will be enabled in the C++ GEMM template, the current implementation for `BF16 xBF16` and `int8xint8` AMX micro-kernel would produce incorrect output. I've highlighted one of the lines of code that would have to be changed.

<img width="666" alt="image" src="https://github.com/user-attachments/assets/77907918-c174-4a52-a36b-b613bf0396a8">

Without weight-prepacking at upper-level, it may have produced correct output, but with weight-prepacking in the GEMM template, it'd produce an incorrect output.

So, if we'd like to support cases in which weight-prepacking would not be done at upper level (but B would have VNNI layout), why are we only focusing on int8 WoQ, and delaying the landing of this PR for something that can't be covered by UTs at this point? Why are we not modifying the existing `BF16xBF16` & `int8xint8` GEMM implementations as well? May I modify their implementation in this PR as well?

Thanks!

jgong5 reviewed (changes requested):
As we discussed offline, please do not assume the B is contiguous.

sanchitintel committed:
Revert to the second implementation

Don't assume weight-packing at GEMM template level

sanchitintel committed:
Add a const variable block_n

Its value would also be known at runtime, so it wouldn't affect performance

leslie-fang-intel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,44 +529,46 @@ class CppMicroGemmAMX(CppMicroGemm):
-                    num_b_rows
-                );
-            }
+        {{input2_t}}* base_addr = const_cast<{{input2_t}}*>(B) + base_idx;
```
The Micro GEMM assumes inputs of B is 2D (K*N) and unpacked. For example, if the original B for the whole template has been packed into [N//Block_N, K, Block_N], then the input of Micro GEMM should be a slice such as [Kc, Block_N].

sanchitintel replied:
> if the original B for the whole template has been packed into [N//Block_N, K, Block_N], then the input of Micro GEMM should be a slice such as [Kc, Block_N]

But if `N` in the micro-kernel wouldn't be equal to `block_n` in the future, the input to the micro-GEMM wouldn't be a `[Kc, block_n]` slice. How would it be sliced then? Thanks!

sanchitintel replied:
That means the weight-packing implementation would change in that case. Thanks for clarifying!

leslie-fang-intel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -514,11 +514,9 @@ class CppMicroGemmAMX(CppMicroGemm):
-    const auto tail_k_size = K - last_k_offset;
     // we cache K * {{block_n}} elements of dequantized B
-    const auto buf_size = K * {{block_n}} * sizeof({{input_t}});
+    const auto block_n = {{block_n}};
```
nit: why not use `{{block_n}}` directly?

sanchitintel replied:
I didn't do that because of style concerns - then only one argument to the API call of the inner-kernel would have not been a variable.

As Jiong mentioned, I'll create another variable & replace argument name `ldb` with it.

leslie-fang-intel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,44 +529,46 @@ class CppMicroGemmAMX(CppMicroGemm):
+                dequantized_B_buf + idx_dq,
+                static_cast<int64_t>(block_n)
+            );
+        {%- elif block_n == 48 %}
```
Add a fallback path when `block_n` is not in [16, 32, 48]?

jgong5 commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,44 +529,46 @@ class CppMicroGemmAMX(CppMicroGemm):
+    for (int64_t n = 0; n < N; n += block_n) {
 {%- if use_cached_dequantized_B %}
-        // Dequantize K * 32 int8 B elements into BF16
+        // Dequantize K * block_n int8 B elements into BF16
```
I guess for all these comments you can directly use `{{block_n}}` instead of `block_n`.

jgong5 commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -590,7 +590,11 @@ class CppMicroGemmAMX(CppMicroGemm):
                     C + m * ldc + n,
                     K,
                     lda,
+{%- if not use_cached_dequantized_B %}
                     ldb,
+{%- else %}
+                    block_n,
+{%- endif %}
```
Shall we define a new variable for the actual `ldb` so that we don't have to do `{% if -else %}` in several places?

sanchitintel replied:
Will do, thanks! We might as well do that for `B` as well.

sanchitintel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -514,11 +514,9 @@ class CppMicroGemmAMX(CppMicroGemm):
-    const auto tail_k_size = K - last_k_offset;
     // we cache K * {{block_n}} elements of dequantized B
-    const auto buf_size = K * {{block_n}} * sizeof({{input_t}});
+    const auto block_n = {{block_n}};
```
Sure, thanks!

sanchitintel replied:
Just curious - had you recommended this change for some reason related to style? I mean it shouldn't have affected performance, as the variable's value would have been known at compile-time.

Thanks

jgong5 replied:
I feel it is also easier to understand the specialized codegen?

sanchitintel commented:
> makes the logic limited to handle 16, 32 and 48

Can we also add a note on how/why a particular set of `[block_m, block_n, block_k]` values were chosen for various dtypes for the AMX micro-kernel in the first place? That may make it easier for someone other than the original authors to understand & experiment with any new configurations. Thanks!

sanchitintel committed:
Revise as per review comments

sanchitintel committed:
Fix typo

leslie-fang-intel commented on torch/_inductor/codegen/cpp_micro_gemm.py:
```diff
@@ -531,44 +528,42 @@ class CppMicroGemmAMX(CppMicroGemm):
+        {%- for vec_idx in range(0, block_n - 1, 32) %}
+            auto b_int8 = at::vec::Vectorized<int8_t>::loadu(
+                base_addr + idx_q + {{vec_idx}} ,
+                static_cast<int64_t>(32)
```
Hi @sanchitintel, will this code work on AVX2 machine?

sanchitintel replied:
AMX is unsupported on machines that support AVX2, but not AVX512.

That's why I didn't ask you a similar question at the time of #136630. :)

sanchitintel committed:
Fix lint

sanchitintel commented:
@pytorchbot rebase -b main
